File: golang/uPIMulator/src/linker/logic/linker_script.go
package logic

import (
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
	"uPIMulator/src/linker/kernel"
	"uPIMulator/src/misc"
)

type LinkerScript struct {
	command_line_parser *misc.CommandLineParser

	num_tasklets           int
	min_access_granularity int64

	linker_constants map[string]*LinkerConstant
}

func (this *LinkerScript) Init(command_line_parser *misc.CommandLineParser) {
	this.command_line_parser = command_line_parser

	this.num_tasklets = int(command_line_parser.IntParameter("num_tasklets"))
	this.min_access_granularity = command_line_parser.IntParameter("min_access_granularity")

	this.linker_constants = make(map[string]*LinkerConstant, 0)

	this.InitLinkerConstants()
}

func (this *LinkerScript) Assign(executable *kernel.Executable) {
	this.AssignAtomic(executable)
	this.AssignIram(executable)
	this.AssignWram(executable)
	this.AssignMram(executable)
}

func (this *LinkerScript) HasLinkerConstant(name string) bool {
	_, found := this.linker_constants[name]
	return found
}

func (this *LinkerScript) LinkerConstant(name string) *LinkerConstant {
	return this.linker_constants[name]
}

func (this *LinkerScript) InitLinkerConstants() {
	config_loader := new(misc.ConfigLoader)
	config_loader.Init()

	max_num_tasklets := config_loader.MaxNumTasklets()
	stack_size := config_loader.StackSize()

	this.linker_constants["NR_TASKLETS"] = new(LinkerConstant)
	this.linker_constants["NR_TASKLETS"].Init("NR_TASKLETS")
	this.linker_constants["NR_TASKLETS"].SetValue(int64(this.num_tasklets))

	for i := 0; i < max_num_tasklets; i++ {
		stack_size_tasklet := "STACK_SIZE_TASKLET_" + strconv.Itoa(i)

		this.linker_constants[stack_size_tasklet] = new(LinkerConstant)
		this.linker_constants[stack_size_tasklet].Init(stack_size_tasklet)
		this.linker_constants[stack_size_tasklet].SetValue(stack_size)
	}

	this.linker_constants["__atomic_start_addr"] = new(LinkerConstant)
	this.linker_constants["__atomic_start_addr"].Init("__atomic_start_addr")

	this.linker_constants["__atomic_used_addr"] = new(LinkerConstant)
	this.linker_constants["__atomic_used_addr"].Init("__atomic_used_addr")

	this.linker_constants["__atomic_end_addr"] = new(LinkerConstant)
	this.linker_constants["__atomic_end_addr"].Init("__atomic_end_addr")

	this.linker_constants["__rodata_start_addr"] = new(LinkerConstant)
	this.linker_constants["__rodata_start_addr"].Init("__rodata_start_addr")

	this.linker_constants["__rodata_end_addr"] = new(LinkerConstant)
	this.linker_constants["__rodata_end_addr"].Init("__rodata_end_addr")

	for i := 0; i < max_num_tasklets; i++ {
		sys_stack_thread := "__sys_stack_thread_" + strconv.Itoa(i)

		this.linker_constants[sys_stack_thread] = new(LinkerConstant)
		this.linker_constants[sys_stack_thread].Init(sys_stack_thread)
	}

	this.linker_constants["__sw_cache_buffer"] = new(LinkerConstant)
	this.linker_constants["__sw_cache_buffer"].Init("__sw_cache_buffer")

	this.linker_constants["__sys_heap_pointer_reset"] = new(LinkerConstant)
	this.linker_constants["__sys_heap_pointer_reset"].Init("__sys_heap_pointer_reset")

	this.linker_constants["__sys_used_mram_end"] = new(LinkerConstant)
	this.linker_constants["__sys_used_mram_end"].Init("__sys_used_mram_end")
}

func (this *LinkerScript) AssignAtomic(executable *kernel.Executable) {
	config_loader := new(misc.ConfigLoader)
	config_loader.Init()

	cur_address := config_loader.AtomicOffset()

	this.linker_constants["__atomic_start_addr"].SetValue(cur_address)

	this.linker_constants["__atomic_used_addr"].SetValue(cur_address)

	for section, _ := range executable.Sections(kernel.ATOMIC) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	this.linker_constants["__atomic_end_addr"].SetValue(cur_address)

	if cur_address >= config_loader.AtomicOffset()+config_loader.AtomicSize() {
		err := errors.New("address is larger than the atomic end address")
		panic(err)
	}
}

func (this *LinkerScript) AssignIram(executable *kernel.Executable) {
	config_loader := new(misc.ConfigLoader)
	config_loader.Init()

	cur_address := config_loader.IramOffset()

	bootstrap := executable.Section(kernel.TEXT, "__bootstrap")
	if bootstrap == nil {
		err := errors.New("bootstrap is not found")
		panic(err)
	}

	bootstrap.SetAddress(cur_address)
	cur_address += bootstrap.Size()

	text_default := executable.Section(kernel.TEXT, "")
	if text_default != nil {
		text_default.SetAddress(cur_address)
		cur_address += text_default.Size()
	}

	for section, _ := range executable.Sections(kernel.TEXT) {
		if section.Name() != "__bootstrap" && section.Name() != "" {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	if cur_address >= config_loader.IramOffset()+config_loader.IramSize() {
		err := errors.New("address is larger than the IRAM end address")
		panic(err)
	}
}

func (this *LinkerScript) AssignWram(executable *kernel.Executable) {
	config_loader := new(misc.ConfigLoader)
	config_loader.Init()

	cur_address := config_loader.WramOffset()

	sys_zero := executable.Section(kernel.DATA, "__sys_zero")
	if sys_zero != nil {
		sys_zero.SetAddress(cur_address)
		cur_address += sys_zero.Size()
	}

	immediate_memory := executable.Section(kernel.DATA, "immediate_memory")
	if immediate_memory != nil {
		immediate_memory.SetAddress(cur_address)
		cur_address += immediate_memory.Size()
	}

	for section, _ := range executable.Sections(kernel.DATA) {
		if strings.Contains(section.Name(), "immediate_memory.") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	this.linker_constants["__rodata_start_addr"].SetValue(cur_address)

	rodata_default := executable.Section(kernel.RODATA, "")
	if rodata_default != nil {
		rodata_default.SetAddress(cur_address)
		cur_address += rodata_default.Size()
	}

	for section, _ := range executable.Sections(kernel.RODATA) {
		if section.Name() != "" {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	this.linker_constants["__rodata_end_addr"].SetValue(cur_address)

	bss_default := executable.Section(kernel.BSS, "")
	if bss_default != nil {
		bss_default.SetAddress(cur_address)
		cur_address += bss_default.Size()
	}

	for section, _ := range executable.Sections(kernel.BSS) {
		if section.Name() != "" {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	sys_keep := executable.Section(kernel.DATA, "__sys_keep")
	if sys_keep != nil {
		sys_keep.SetAddress(cur_address)
		cur_address += sys_keep.Size()
	}

	data_default := executable.Section(kernel.DATA, "")
	if data_default != nil {
		data_default.SetAddress(cur_address)
		cur_address += data_default.Size()
	}

	for section, _ := range executable.Sections(kernel.DATA) {
		if section.Name() != "__sys_zero" &&
			section.Name() != "__sys_keep" &&
			!strings.Contains(section.Name(), "immediate_memory") &&
			section.Name() != "" {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	// TODO(bongjoon.hyun@gmail.com): figure out ".data.__sys_host" section

	dpu_host := executable.Section(kernel.DPU_HOST, "")
	if dpu_host != nil {
		dpu_host.SetAddress(cur_address)
		cur_address += dpu_host.Size()
	}

	// TODO(bongjoon.hyun@gmail.com): figure out ".data.__sys_profilng" section

	// TODO(bongjoon.hyun@gmail.com): figure out ".data.stacks" section

	for i := 0; i < config_loader.MaxNumTasklets(); i++ {
		sys_stack_thread := "__sys_stack_thread_" + strconv.Itoa(i)
		this.linker_constants[sys_stack_thread].SetValue(cur_address)

		stack_size_tasklet := "STACK_SIZE_TASKLET_" + strconv.Itoa(i)
		cur_address += this.linker_constants[stack_size_tasklet].Value()
	}

	// TODO(bongjoon.hyun@gmail.com): figure out ".data.sw_cache" section

	this.linker_constants["__sw_cache_buffer"].SetValue(cur_address)
	cur_address += int64(8 * config_loader.MaxNumTasklets())

	// TODO(bongjoon.hyun@gmail.com): figure out ".data.heap_pointer_reset" section

	cur_address = int64(
		math.Ceil(float64(cur_address)/float64(this.min_access_granularity)),
	) * int64(
		this.min_access_granularity,
	)
	this.linker_constants["__sys_heap_pointer_reset"].SetValue(cur_address)

	if cur_address >= config_loader.WramOffset()+config_loader.WramSize() {
		err := errors.New("address is larger than the WRAM end address")
		panic(err)
	}
}

func (this *LinkerScript) AssignMram(executable *kernel.Executable) {
	config_loader := new(misc.ConfigLoader)
	config_loader.Init()

	cur_address := config_loader.MramOffset()

	for section, _ := range executable.Sections(kernel.DEBUG_ABBREV) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.DEBUG_FRAME) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.DEBUG_INFO) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.DEBUG_LINE) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.DEBUG_LOC) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.DEBUG_RANGES) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.DEBUG_STR) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	for section, _ := range executable.Sections(kernel.STACK_SIZES) {
		section.SetAddress(cur_address)
		cur_address += section.Size()
	}

	noinit := executable.Section(kernel.MRAM, "noinit")
	if noinit != nil {
		noinit.SetAddress(cur_address)
		cur_address += noinit.Size()
	}

	for section, _ := range executable.Sections(kernel.MRAM) {
		if strings.Contains(section.Name(), "noinit.") &&
			!strings.Contains(section.Name(), "noinit.keep") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	noinit_keep := executable.Section(kernel.MRAM, "noinit.keep")
	if noinit_keep != nil {
		noinit_keep.SetAddress(cur_address)
		cur_address += noinit_keep.Size()
	}

	for section, _ := range executable.Sections(kernel.MRAM) {
		if strings.Contains(section.Name(), "noinit.keep") &&
			!strings.Contains(section.Name(), "noinit.keep.") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	for section, _ := range executable.Sections(kernel.MRAM) {
		if strings.Contains(section.Name(), "noinit.keep.") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	mram_default := executable.Section(kernel.MRAM, "")
	if mram_default != nil {
		mram_default.SetAddress(cur_address)
		cur_address += mram_default.Size()
	}

	for section, _ := range executable.Sections(kernel.MRAM) {
		if section.Name() != "" &&
			!strings.Contains(section.Name(), "noinit") &&
			!strings.Contains(section.Name(), "keep") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	for section, _ := range executable.Sections(kernel.MRAM) {
		if strings.Contains(section.Name(), "keep") &&
			!strings.Contains(section.Name(), "keep.") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	for section, _ := range executable.Sections(kernel.MRAM) {
		if strings.Contains(section.Name(), "keep.") {
			section.SetAddress(cur_address)
			cur_address += section.Size()
		}
	}

	cur_address = int64(
		math.Ceil(float64(cur_address)/float64(this.min_access_granularity)),
	) * this.min_access_granularity

	this.linker_constants["__sys_used_mram_end"].SetValue(cur_address)

	if cur_address >= config_loader.MramOffset()+config_loader.MramSize() {
		err := errors.New("address is larger than the MRAM end address")
		panic(err)
	}
}

func (this *LinkerScript) DumpValues(path string) {
	lines := make([]string, 0)

	for _, linker_constant := range this.linker_constants {
		line := fmt.Sprintf("%s: %d", linker_constant.Name(), linker_constant.Value())
		lines = append(lines, line)
	}

	file_dumper := new(misc.FileDumper)
	file_dumper.Init(path)
	file_dumper.WriteLines(lines)
}

File: python_cpp/uPIMulator_frontend/src/linker_/linker_script.py
import math
from typing import Dict, Set

from abi.binary.executable import Executable
from abi.label.label import Label
from abi.section.section_name import SectionName
from abi.word.instruction_word import InstructionWord
from util.config_loader import ConfigLoader


class LinkerScript:
    def __init__(self, num_tasklets: int):
        self._num_tasklets: int = num_tasklets

        self._constants: Dict[str, int] = self._init_constants()
        self._symbols: Set[Label] = self._init_symbols()

    def constant(self, constant_name: str) -> int:
        return self._constants[constant_name]

    def constant_names(self) -> Set[str]:
        return set(self._constants.keys())

    def symbol(self, symbol_name: str) -> Label:
        for symbol in self._symbols:
            if symbol.name() == symbol_name:
                return symbol
        raise ValueError

    def symbol_names(self) -> Set[str]:
        return {symbol.name() for symbol in self._symbols}

    def assign_address(self, executable: Executable) -> None:
        self._assign_iram(executable)
        self._assign_atomic(executable)
        self._assign_wram(executable)
        self._assign_mram(executable)

    def _assign_iram(self, executable: Executable) -> None:
        cur_address = ConfigLoader.iram_offset()
        bootstrap = executable.section(SectionName.TEXT, "__bootstrap")

        assert bootstrap is not None
        assert bootstrap.address() is None

        bootstrap.set_address(cur_address)
        cur_address += bootstrap.size()
        assert bootstrap.size() % InstructionWord().size() == 0

        text_section = executable.section(SectionName.TEXT, "")
        if text_section is not None:
            assert text_section.address() is None

            text_section.set_address(cur_address)
            cur_address += text_section.size()

            assert text_section.size() % InstructionWord().size() == 0

        for section in executable.sections(SectionName.TEXT):
            if section.name() != "__bootstrap" and section.name() != "":
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

                assert section.size() % InstructionWord().size() == 0

        assert cur_address < (ConfigLoader.iram_offset() + ConfigLoader.iram_size())

    def _assign_atomic(self, executable: Executable) -> None:
        cur_address = ConfigLoader.atomic_offset()
        self.symbol("__atomic_start_addr").set_address(cur_address)

        # TODO(bongjoon.hyun@gmail.com): the original UPMEM linker script adds 200 to the cur_address
        # cur_address += 200

        self.symbol("__atomic_used_addr").set_address(cur_address)
        for section in executable.sections(SectionName.ATOMIC):
            assert section.address() is None

            section.set_address(cur_address)
            cur_address += section.size()

        self.symbol("__atomic_end_addr").set_address(cur_address)

        assert cur_address < (ConfigLoader.atomic_offset() + ConfigLoader.atomic_size())

    def _assign_wram(self, executable: Executable) -> None:
        cur_address = ConfigLoader.wram_offset()

        sys_zero = executable.section(SectionName.DATA, "__sys_zero")
        if sys_zero is not None:
            assert sys_zero.address() is None

            sys_zero.set_address(cur_address)
            cur_address += sys_zero.size()

        immediate_memory = executable.section(SectionName.DATA, "immediate_memory")
        if immediate_memory is not None:
            assert immediate_memory.address() is None

            immediate_memory.set_address(cur_address)
            cur_address += immediate_memory.size()

        for section in executable.sections(SectionName.DATA):
            if "immediate_memory." in section.name():
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

        self.symbol("__rodata_start_addr").set_address(cur_address)
        rodata_section = executable.section(SectionName.RODATA, "")
        if rodata_section is not None:
            assert rodata_section.address() is None

            rodata_section.set_address(cur_address)
            cur_address += rodata_section.size()

        for section in executable.sections(SectionName.RODATA):
            if section.name() != "":
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()
        self.symbol("__rodata_end_addr").set_address(cur_address)

        bss_section = executable.section(SectionName.BSS, "")
        if bss_section is not None:
            assert bss_section.address() is None

            bss_section.set_address(cur_address)
            cur_address += bss_section.size()

        for section in executable.sections(SectionName.BSS):
            if section.name() != "":
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

        sys_keep = executable.section(SectionName.DATA, "__sys_keep")
        if sys_keep is not None:
            assert sys_keep.address() is None

            sys_keep.set_address(cur_address)
            cur_address += sys_keep.size()

        data_section = executable.section(SectionName.DATA, "")
        if data_section is not None:
            assert data_section.address() is None

            data_section.set_address(cur_address)
            cur_address += data_section.size()

        for section in executable.sections(SectionName.DATA):
            if (
                section.name() != "__sys_zero"
                and section.name() != "__sys_keep"
                and section.name() != ""
                and "immediate_memory" not in section.name()
            ):
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

        dpu_host = executable.section(SectionName.DPU_HOST, "")
        if dpu_host is not None:
            assert dpu_host.address() is None

            dpu_host.set_address(cur_address)
            cur_address += dpu_host.size()

        for i in range(ConfigLoader.max_num_tasklets()):
            self.symbol(f"__sys_stack_thread_{i}").set_address(cur_address)
            cur_address += self.constant(f"STACK_SIZE_TASKLET_{i}")

        self.symbol("__sw_cache_buffer").set_address(cur_address)
        cur_address += 8 * self._num_tasklets

        cur_address = (
            math.ceil(cur_address / ConfigLoader.min_access_granularity()) * ConfigLoader.min_access_granularity()
        )
        self.symbol("__sys_heap_pointer_reset").set_address(cur_address)

        assert cur_address < (ConfigLoader.wram_offset() + ConfigLoader.wram_size())

    def _assign_mram(self, executable: Executable) -> None:
        cur_address = ConfigLoader.mram_offset()

        debug_abbrev_section = executable.section(SectionName.DEBUG_ABBREV, "")
        if debug_abbrev_section is not None:
            assert debug_abbrev_section.address() is None

            debug_abbrev_section.set_address(cur_address)
            cur_address += debug_abbrev_section.size()

        debug_frame_section = executable.section(SectionName.DEBUG_FRAME, "")
        if debug_frame_section is not None:
            assert debug_frame_section.address() is None

            debug_frame_section.set_address(cur_address)
            cur_address += debug_frame_section.size()

        debug_info_section = executable.section(SectionName.DEBUG_INFO, "")
        if debug_info_section is not None:
            assert debug_info_section.address() is None

            debug_info_section.set_address(cur_address)
            cur_address += debug_info_section.size()

        debug_line_section = executable.section(SectionName.DEBUG_LINE, "")
        if debug_line_section is not None:
            assert debug_line_section.address() is None

            debug_line_section.set_address(cur_address)
            cur_address += debug_line_section.size()

        debug_loc_section = executable.section(SectionName.DEBUG_LOC, "")
        if debug_loc_section is not None:
            assert debug_loc_section.address() is None

            debug_loc_section.set_address(cur_address)
            cur_address += debug_loc_section.size()

        debug_ranges_section = executable.section(SectionName.DEBUG_RANGES, "")
        if debug_ranges_section is not None:
            assert debug_ranges_section.address() is None

            debug_ranges_section.set_address(cur_address)
            cur_address += debug_ranges_section.size()

        debug_str_section = executable.section(SectionName.DEBUG_STR, "")
        if debug_str_section is not None:
            assert debug_str_section.address() is None

            debug_str_section.set_address(cur_address)
            cur_address += debug_str_section.size()

        stack_sizes_section = executable.section(SectionName.STACK_SIZES, "")
        if stack_sizes_section is not None:
            assert stack_sizes_section.address() is None

            stack_sizes_section.set_address(cur_address)
            cur_address += stack_sizes_section.size()

        noinit = executable.section(SectionName.MRAM, "noinit")
        if noinit is not None:
            assert noinit.address() is None

            noinit.set_address(cur_address)
            cur_address += noinit.size()

        for section in executable.sections(SectionName.MRAM):
            if "noinit." in section.name():
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

        mram_section = executable.section(SectionName.MRAM, "")
        if mram_section is not None:
            assert mram_section.address() is None

            mram_section.set_address(cur_address)
            cur_address += mram_section.size()

        for section in executable.sections(SectionName.MRAM):
            if section.name() != "" and "noinit" not in section.name() and "keep" not in section.name():
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

        keep = executable.section(SectionName.MRAM, "keep")
        if keep is not None:
            assert keep.address() is None

            keep.set_address(cur_address)
            cur_address += keep.size()

        for section in executable.sections(SectionName.MRAM):
            if "keep." in section.name():
                assert section.address() is None

                section.set_address(cur_address)
                cur_address += section.size()

        cur_address = (
            math.ceil(cur_address / ConfigLoader.min_access_granularity()) * ConfigLoader.min_access_granularity()
        )
        self.symbol("__sys_used_mram_end").set_address(cur_address)

        assert cur_address < (ConfigLoader.mram_offset() + ConfigLoader.mram_size())

    def _init_constants(self) -> Dict[str, int]:
        return {
            "NR_TASKLETS": self._num_tasklets,
            **self._init_stack_size_constants(),
        }

    def _init_stack_size_constants(self) -> Dict[str, int]:
        return {f"STACK_SIZE_TASKLET_{i}": ConfigLoader.stack_size() for i in range(ConfigLoader.max_num_tasklets())}

    def _init_symbols(self) -> Set[Label]:
        return {
            *self._init_atomic_symbols(),
            *self._init_wram_symbols(),
            *self._init_mram_symbols(),
        }

    def _init_atomic_symbols(self) -> Set[Label]:
        return {
            Label("__atomic_start_addr"),
            Label("__atomic_used_addr"),
            Label("__atomic_end_addr"),
        }

    def _init_wram_symbols(self) -> Set[Label]:
        return {
            Label("__rodata_start_addr"),
            Label("__rodata_end_addr"),
            *self._init_stack_symbols(),
            Label("__sw_cache_buffer"),
            Label("__sys_heap_pointer_reset"),
        }

    def _init_stack_symbols(self) -> Set[Label]:
        return {Label(f"__sys_stack_thread_{i}") for i in range(ConfigLoader.max_num_tasklets())}

    def _init_mram_symbols(self) -> Set[Label]:
        return {Label("__sys_used_mram_end")}
